pkg/controllers/federatedhpa/centralizedhpa_controller.go

```diff
@@ -0,0 +1,1283 @@
+package federatedhpa
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	autoscalingv2 "k8s.io/api/autoscaling/v2"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	listcorev1 "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	autoscalingv1alpha1 "github.com/karmada-io/karmada/pkg/apis/autoscaling/v1alpha1"
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+	policyv1alpha1 "github.com/karmada-io/karmada/pkg/apis/policy/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+	"github.com/karmada-io/karmada/pkg/controllers/federatedhpa/monitor"
+	"github.com/karmada-io/karmada/pkg/sharedcli/ratelimiterflag"
+	"github.com/karmada-io/karmada/pkg/util"
+	"github.com/karmada-io/karmada/pkg/util/fedinformer/typedmanager"
+	"github.com/karmada-io/karmada/pkg/util/helper"
+	"github.com/karmada-io/karmada/pkg/util/lifted/selectors"
+)
+
+// CentralizedHpa-controller is borrowed from the HPA controller of Kubernetes.
+// The referenced code has been marked in the comment.
+
+// ControllerName is the controller name that will be used when reporting events.
+const ControllerName = "federatedHPA-controller"
+
+var (
+	podGVR              = corev1.SchemeGroupVersion.WithResource("pods")
+	scaleUpLimitFactor  = 2.0
+	scaleUpLimitMinimum = 4.0
+)
+
+var (
+	// errSpec is used to determine if the error comes from the spec of HPA object in reconcileAutoscaler.
+	// All such errors should have this error as a root error so that the upstream function can distinguish spec errors from internal errors.
+	// e.g., fmt.Errorf("invalid spec%w", errSpec)
+	errSpec error = errors.New("")
+)
+
+// CentralizedHorizontalPodAutoscalerController is to sync HorizontalPodAutoscaler.
+type CentralizedHorizontalPodAutoscalerController struct {
```

Comment:
Still `Centralized`?
Given this struct is defined under the package `federatedhpa`, we can short the name with `HPAController` or `FederatedHPAController`.